core/src/main/java/com/graphhopper/routing/weighting/custom/ExpressionBuilder.java

```diff
@@ -0,0 +1,382 @@
+/*
+ *  Licensed to GraphHopper GmbH under one or more contributor
+ *  license agreements. See the NOTICE file distributed with this work for
+ *  additional information regarding copyright ownership.
+ *
+ *  GraphHopper GmbH licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except in
+ *  compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package com.graphhopper.routing.weighting.custom;
+
+import com.graphhopper.json.geo.JsonFeature;
+import com.graphhopper.routing.ev.DecimalEncodedValue;
+import com.graphhopper.routing.ev.EncodedValue;
+import com.graphhopper.routing.ev.EncodedValueLookup;
+import com.graphhopper.routing.util.CustomModel;
+import com.graphhopper.routing.util.EncodingManager;
+import com.graphhopper.util.EdgeIteratorState;
+import com.graphhopper.util.GHUtility;
+import com.graphhopper.util.shapes.BBox;
+import com.graphhopper.util.shapes.Polygon;
+import org.codehaus.commons.compiler.CompileException;
+import org.codehaus.commons.compiler.Location;
+import org.codehaus.commons.compiler.io.Readers;
+import org.codehaus.janino.Scanner;
+import org.codehaus.janino.*;
+import org.codehaus.janino.util.DeepCopier;
+import org.locationtech.jts.geom.prep.PreparedGeometryFactory;
+
+import java.io.*;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicLong;
+
+class ExpressionBuilder {
```

Comment:
Sure, this evolved a lot and can call it differently. The only reason for this class was to avoid "polluting" CustomWeighting

Reply:
Should we name CustomWeightingHelper something like SpeedAndPriorityProvider and ExpressionBuilder then SpeedAndPriorityProviderBuilder? (Hmmh, don't like it too much :D )

Reply:
Yeah I do not like `SpeedAndPriorityProviderBuilder` either. Especially the `XAndY` part. We could call it `CustomModelInterpreter` maybe? Basically it takes a custom model and returns something we can use to calculate a speed and priority for an edge. Actually I wish this were two methods: one for speed and one for priority, but that probably does not work because we need janino to create one class that does both and otherwise we would have to create two classes which would be worse for performance?

Reply:
I'm not sure if it would be worse for performance but we would need two classes, two caches, two init calls and more. I currently prefer the variation you have in your PR where the interface uses two separate interfaces but our janino implementation can remain via one class.

Reply:
Ok that makes sense. Another thing that came to my mind: Could we limit Janino to parsing boolean expressions entirely? I mean instead of creating one subclass per custom weighting create one subclass per boolean expression and cache all these expressions (and share between custom weightings). Instead of functions like (edge, reverse)->double for speed, priority etc. we would only get functions like (edge, reverse)->boolean and everything else would be plain Java (without Janino). I did not really look at details about Janino and caching, so this is just a thought...

Reply:
In general this is possible as Janino allows parsing also only single expressions, but we would need to the set up every single expression manually (like telling janino where to get the constants via imports and the variables etc). So this will not only be very tedious, but might also have a negative impact on speed due to this overhead of "context switch".

Reply:
Actually we already parse every single expression but it would not compile (missing context like imports and encoded values etc) and so we would need to do this on every call. This would be also limiting as we would know the number of operators and currently we could theoretically even do `a * b > 10` or call (whitelisted) methods or Math operations etc.

Reply:
Hm I don't understand this yet. Yes the Janino class needs all the imports and constants to parse the different possible expressions, but for the current CustomWeightingHelperSubclass its the same? Why do we need to setup the single expressions manually? I meant we could simply feed the expressions (the current keys of the custom model) to Janino and build a cache containing these. So for example if multiple custom weightings use `road_class == PRIMARY` it will be shared between these custom weightings. But I am not sure if this is even 'better' or faster. To me it just seems simpler and easier to think about that there is some parser that converts boolean expressions to compiled (edge,reverse)->boolean functions than to think of a parser that knows about all the rules of the custom model language.

Reply:
> I meant we could simply feed the expressions (the current keys of the custom model) to Janino and build a cache containing these

Parsed statements still needs to be compiled. So only because I know that e.g. `road_class` is some token we would have to tell the compiled expression the value of `road_class` so that janino then evaluates the boolean expression for us. E.g. for: `road_class == PRIMARY` this is something like
```
loop over compiledExpressions:
  compiledExpression.setArgument(0, edge.get(road_class_enc)) // loop over all arguments
  if(compiledExpression.evaluate()) // continue or break depending on the operator (`&&` or `||`) or more complicated if there are `()` brakets etc.
```

So basically we would do part of the job why we introduced Janino :)

> than to think of a parser that knows about all the rules of the custom model language.

I think of Janino as some Java code that calculates the speed and the priority based on some simple if-clauses. E.g. you can always print the create Java code via
```
StringWriter sw = new StringWriter();
Unparser.unparse(cu, sw);
```

Reply:
> I think of Janino as some Java code that calculates the speed and the priority based on some simple if-clauses.

But Janino can only understand the parts of custom_model.yml that are Java right? To me it looks like we create a Janino subclass that knows about the structure of custom_model.yml (our custom model language) such that it can map edges to speeds and priorities. I would have expected we use Janino to map edges to booleans (only parse the Java parts of custom_model.yml) and then apply our custom model language using these boolean functions. But maybe I have some misunderstanding here and I will have a closer look at Janino to understand it better.

Reply:
Let's say we have this custom model:

```yaml
speed:
   - if: road_class == PRIMARY && surface == GRAVEL
     factor: 0.5
   - if: road_environment == TUNNEL
     factor: 0.9
```

I thought we could create two subclasses of some interface like `EdgeToBooleanMapping` with janino and then use this to build our custom language around this like this:

```java
List<EdgeToBooleanMapping> mappings = new ArrayList<>();
for (Entry entry : customModel.getSpeed().getEntries()) {
    mappings.add(janino.compileOrLoadFromCache(createJaninoSubclass(entry.get("if"), entry.get("factor"));
}
double speed = 1;
for (EdgeToBooleanMapping mapping : mappings) {
    speed *= mapping.get(edge, reverse);
}
```

`createJaninoSubclass` would work exactly the same way as the code that currently creates a subclass of `CustomWeightingHelper` (same imports and context), except that it produces something that returns a boolean for a single expression instead of a speed or priority.

So this would work similar to what we do right now, doesn't it?

But I am not sure if this is even better, because in this case we need to compile multiple subclasses for every custom weighting. But on the other hand we can share some of them between custom weightings. Also the code above would only be compiled once and run for all custom weightings (not sure if this is better for JIT optimization).

Reply:
Ok, I see.

> But I am not sure if this is even better

I'm also unsure if this is better. Please note that we currently fetch e.g. `road_class` once in getSpeed and can then use it in all of the following expressions. If the expressions are separate we would need some local cache (`Map`) for this or fetch this value multiple times (or introduce the cache inside the `BaseGraph`). So both is currently more work compared to only using a variable inside the same method. Also the for-loops are currently not existent, so this would also introduce slightly more work. 

Still it would reduce the complexity a bit as we only need to compile the previously scanned expressions and we won't need DeepCopier and probably also no SimpleCompiler, but IMO this is only an implementation detail (for now).

Reply:
> Note that we currently fetch e.g. road_class once in getSpeed and can then use it in all of the following expressions. 

Yes good point, this might be important. What exactly does `fetching` mean here? The flags are cached in the EdgeIteratorState anyway, right?

> Also the for-loops are currently not existent, so this would also introduce slightly more work

Ok just the fact that you already did it the way it is now is reason enough to keep it as it is for now. Especially, because so far we can only speculate which approach is faster and if this even matters. But I am pretty sure it would be a bit cleaner code (unless we need some complicated workarounds to achieve the same performance).

> , but IMO this is only an implementation detail (for now).

Yes, totally. So let's just keep this in mind and leave as is for now.

Reply:
> What exactly does `fetching` mean here?

Yes, flags are cached but "extracting" the int or double number from the flags via bit shifting is a tiny overhead. And probably the bigger overhead is to set the arguments and get the result and pipe this through the other expressions.

> But I am pretty sure it would be a bit cleaner code (unless we need some complicated workarounds to achieve the same performance)

ok, yes